api/src/controllers/login.js

```diff
@@ -229,10 +251,64 @@ const getBranding = () => {
     });
 };
 
+const renderTokenLogin = (req, res) => {
+  return getBranding()
+    .then(branding => render('tokenLogin', req, branding, { tokenUrl: req.url }))
+    .then(body => res.send(body));
+};
+
+const createSessionRetry = (req, retry= 10) => {
+  return createSession(req).then(sessionRes => {
+    if (sessionRes.statusCode === 200) {
+      return sessionRes;
+    }
+
+    if (retry) {
+      return createSessionRetry(req, retry - 1);
+    }
+
+    throw { status: 449, message: 'Login failed after 10 retries' };
+  });
+};
+
+/**
+ * Generates a session cookie for a user identified by supplied token and hash request params.
+ * The user's password is reset in the process.
+ */
+const tokenLogin = (req, res) => {
+  if (!req.params || !req.params.token || !req.params.hash) {
+    return res.status(400).json({ error: 'Missing required params' });
+  }
+
+  return users
+    .getUserByToken(req.params.token, req.params.hash)
+    .then(userId => {
+      if (!userId) {
+        return res.status(401).json({ error: 'Token invalid / expired' });
+      }
+
+      return users.resetPassword(userId).then(({ user, password }) => {
+        req.body = { user, password };
```

Comment:
Is this going to end up in the haproxy log? I think it's ok because the incoming request will be logged before it gets here and the outgoing logging doesn't log the req body, but it's worth checking.

Reply:
It will be just like any other login request (I initially called `module.exports.post()` with this). 
So if "regular" login requests log the body in haproxy, then this one will too. If they don't ...